`wg-blimp` (Whole Genome BisuLfIte sequencing Methylation analysis Pipeline) can be utilised to analyse WGBS data. It performs alignment, qc, methylation calling, DMR calling and DMR annotation using a multitude of tools.

## Requirements
To run `wg-blimp` you need a UNIX environment that contains an installation of [Bioconda](http://bioconda.github.io/).

## Installation

### Recommended
It is advised to install `wg-blimp` through Bioconda, using:
```
conda install wg-blimp
```

### From source
You can also install `wg-blimp` from source using
```
python setup.py install
```
Using this installation method requires you to make sure all external tools are installed (such as bwa-meth).

## Running wg-blimp

### WGBS pipeline

`wg-blimp` is a cli wrapper for the WGBS pipeline implemented using [Snakemake](http://snakemake.readthedocs.io/). In general, a pipeline config us fed to the Snakemake workflow and the corresponding tools are called. However, `wg-blimp` also provides some commands to ease creation of config files, or working without config files altogether.

The command `wg-blimp run-snakemake` will run the pipeline with its default parameters. Make sure to set the `--cores` and `--genome-build` options appropriately. This command will also internally create a `config.yaml` file containing all parameters used for the analysis.

However, in case the default configurations are not sufficient, users can provide their own configurations. The commands `wg-blimp create-config` and `wg-blimp run-snakemake-from-config` can be used for this purpose.

The folder structure created by `wg-blimp run-snakemake` will look as follows:

* alignment - contains all bam/bai files
* dmr - contains dmr files by different callers
* logs - each pipeline step deposits its logs here
* methylation - methylation bedgraph files
* qc - multiqc and other qc related files
* raw - text files describing which fastq files have been used for each sample
* config.yaml - configuration file used for the analysis

### Shiny GUI

You can use the command `wg-blimp run-shiny` to load one or more project folders (that have been generated by the Snakemake pipeline) into a shiny GUI for easier access.

## Example

Some example `.fastq`can be found on [Sciebo](https://uni-muenster.sciebo.de/s/8Z5uBWhlkunXfbY). You can use the command
```
wg-blimp run-snakemake <folder-with-fastqs> <reference.fa> simulated1,simulated2 simulated3,simulated4 <output-folder>
```

Please not that the pipeline commands also allow a `--use-sample-files` option so sample groups can be loaded from text files instead of comma separates files.


## Config parameters

The following entries are used for running the Snakemake pipeline and may be specified in the `config.yaml` files:

| Key | Value |
| --- | ----- |
| *annotation_allowed_biotypes* | Only genes with this biotype will be annotated in the DMR table |
| *annotation_min_mapq* | When annotating coverage, only use reads with a minimum mapping quality |
| *cgi_annotation_file* | Gzipped csv file used for cg island annotation. |
| *computing_threads* | Number of processors a single job is allowed to use. Remember to use `--cores` parameter for Snakemake. |
| *dmr_tools* | Tools to use for DMR calling. Available: `bsseq`, `camel`, `metilene`
| *gene_annotation_file* | File used for genetic annotation. |
| *group1* | Samples in first group for DMR analysis |
| *group2* | Samples in second group for DMR analysis |
| *io_threads* | IO intensive tools virtually reserve this many cores (while actually using only one) to reduce file system IO load. |
| *min_cov* | Minimum average coverage for methylation calling |
| *min_cpg* | Minimum number of CpGs in a DMR to be called |
| *min_diff* | Minimum average difference between the two groups for DMR calling |
| *output_dir* | Directory containing all files created by the pipeline |
| *promoter_tss_distances* | Distance interval around TSS's to be recognized as promoters in DMR annotation. |
| *rawdir* | Directory containing .fastq files |
| *ref* | .fasta reference file |
| *repeat_masker_annotation_file* | File containing repeat masker annotation |
| *repeat_masker_links* | Repeat masker files are relatively big and are only downloaded on demand from the links specified here. |
| *samples* | All samples (usually concatenation of group1 and group2) |
| *target_files* | Files to be generated by the Snakemake workflow |
| *transcript_start_site_file* | File to read transcription start sites from (for DMR annotation) |

## Reporting errors / Requesting features
If anything goes wrong using `wg-blimp` or any features are missing, feel free to open an issue or to contact Marius WÃ¶ste ( mar.w@wwu.de )
